File: Backend/SER.Tools/Types/Types/Results/Result.cs
using System.Text.Json.Serialization;

namespace SER.Tools.Types.Results;

public class Result
{
	public Error[] Errors { get; }

	[JsonIgnore]
	public Boolean IsSuccess => Errors.Length == 0;

	public Result(List<Error> errors)
	{
		Errors = errors.ToArray();
	}

	[JsonConstructor]
	public Result(Error[] errors)
	{
		Errors = errors;
	}

	public static Result Success()
	{
		return new Result(new List<Error>());
	}

	public static Result Fail(Error error)
	{
		return new Result(new List<Error>() { error });
	}

	public static Result Fail(String error)
	{
		return new Result(new List<Error>() { new Error(error) });
	}
}

File: Frontend/students-employment-records-site/src/tools/result.ts
export class Result {
    constructor(public errors: Error[]) {}

    public isSuccess = this.errors.length === 0

    public get getError() {
        return this.errors[0].message
    }

    public static success<T>(): Result {
        return new Result([])
    }

    public static fail(errors: Error[]): Result {
        return new Result(errors)
    }

    public static fromAny(result: any): Result {
        return new Result(Error.fromAny(result.errors))
    }
}

class Error {
    constructor(public key: string | null, public message: string) {}

    public static fromAny(errors: any[]): Error[] {
        return errors.map((error) => new Error(error.key ?? null, error.message))
    }
}
